File: apps/ios/ArgusKit/Sources/ArgusKit/Realtime/StreamClient.swift
import Foundation
// SocketIO hasn't adopted Sendable; @preconcurrency relaxes strict
// checking at its API boundary only (our own code stays fully checked).
@preconcurrency import SocketIO

// Realtime layer — the Swift counterpart of `apps/web/src/lib/ws.ts`,
// against the event map in `packages/shared-types/src/ws.ts`.
//
// Auth: the server (`stream.gateway.ts#handleConnection`) reads the JWT
// from `handshake.auth.token`. We send it via the Socket.IO CONNECT
// payload (`connect(withPayload:)`) — unlike a header this travels inside
// the Engine.IO CONNECT packet, so it survives proxies.
//
// Delivery is best-effort: on `.connected` after a drop, the app must
// backfill missed chunks over REST (`getSessionChunks(afterSeq:)`) —
// there is no socket-side replay buffer.

/// Small payloads for events that don't carry a full DTO.
public struct IdStatusPayload: Decodable, Equatable, Sendable {
    public let id: String
    public let status: String
}

public struct IdPayload: Decodable, Equatable, Sendable {
    public let id: String
}

public struct SessionCloneFailedPayload: Decodable, Equatable, Sendable {
    public let sessionId: String
    public let reason: String
}

/// fs/git nudges are scoped to the `(machineId, workingDir)` project the
/// runner emitted them for; panels match on that pair. Both fields
/// optional for tolerance — a dropped identity must degrade matching,
/// never decoding.
public struct FSChangedPayload: Decodable, Equatable, Sendable {
    public let path: String
    public let machineId: String?
    public let workingDir: String?
}

public struct GitChangedPayload: Decodable, Equatable, Sendable {
    public let machineId: String?
    public let workingDir: String?
}

/// The server's retention window for an ended task elapsed (or a user
/// dismissed it) — drop the row.
public struct BackgroundTaskRemovedPayload: Decodable, Equatable, Sendable {
    public let machineId: String
    public let workingDir: String
    public let taskId: String
}

/// One chunk of PTY output. `data` is base64 raw bytes; `seq` is the
/// duplicate guard (feed strictly increasing seqs only).
public struct TerminalOutputPayload: Decodable, Equatable, Sendable {
    public let terminalId: String
    public let seq: Int
    public let data: String
}

public struct TerminalClosedPayload: Decodable, Equatable, Sendable {
    public let terminalId: String
    public let exitCode: Int?
    public let reason: String?
}

/// One live event from the `/stream` namespace — the Phase-1 subset plus
/// fleet upkeep. Terminal, background-task, and sidecar-update events are
/// wired in later phases.
public enum ServerEvent: Sendable {
    case connected
    case disconnected
    case socketError(String)

    case chunk(ResultChunk)
    case commandCreated(CommandDTO)
    case commandUpdated(CommandDTO)

    case sessionCreated(SessionDTO)
    case sessionUpdated(SessionDTO)
    case sessionStatus(SessionStatusEvent)
    case sessionCloneFailed(SessionCloneFailedPayload)

    case machineUpsert(MachineDTO)
    case machineStatus(IdStatusPayload)
    case machineRemoved(IdPayload)

    case projectUpsert(ProjectDTO)

    case fsChanged(FSChangedPayload)
    case gitChanged(GitChangedPayload)

    /// Scoped to `project:<machineId>:<workingDir>` rooms
    /// (`subscribe:project`).
    case backgroundTaskUpdated(BackgroundTaskDTO)
    case backgroundTaskRemoved(BackgroundTaskRemovedPayload)

    /// created/updated arrive on the user room; output/closed on the
    /// `terminal:{id}` room (`subscribe:terminal`).
    case terminalCreated(TerminalDTO)
    case terminalUpdated(TerminalDTO)
    case terminalOutput(TerminalOutputPayload)
    case terminalClosed(TerminalClosedPayload)
}

/// Identifies one project room. A Hashable struct rather than a joined
/// string so there's no separator for a `workingDir` to collide on —
/// `("a", "b/c")` and `("a/b", "c")` are distinct rooms.
///
/// File scope, not nested in `StreamClient`: a type nested inside a
/// `@MainActor` class inherits that isolation, which fights the
/// `Sendable` conformance a dictionary key wants.
struct ProjectRoomKey: Hashable, Sendable {
    let machineId: String
    let workingDir: String
}

/// Owns the Socket.IO connection and surfaces typed events as an
/// `AsyncStream`. Create one per login; `connect` after login, `shutdown`
/// on logout.
@MainActor
public final class StreamClient {
    /// Consume once; delivers every event for the client's lifetime.
    public let events: AsyncStream<ServerEvent>
    private let continuation: AsyncStream<ServerEvent>.Continuation

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    public init() {
        var streamContinuation: AsyncStream<ServerEvent>.Continuation!
        self.events = AsyncStream(bufferingPolicy: .unbounded) { streamContinuation = $0 }
        self.continuation = streamContinuation
    }

    // MARK: Lifecycle

    public func connect(baseURL: URL, token: String) {
        shutdownSocketKeepingStream()
        // Reconnection mirrors ws.ts: always retry, capped backoff.
        let manager = SocketManager(
            socketURL: baseURL,
            config: [
                .forceWebsockets(true),
                .reconnects(true),
                .reconnectWait(1),
                .reconnectWaitMax(10),
                .log(false),
            ]
        )
        let socket = manager.socket(forNamespace: "/stream")
        self.manager = manager
        self.socket = socket
        registerHandlers(on: socket)
        socket.connect(withPayload: ["token": token])
    }

    public func shutdown() {
        shutdownSocketKeepingStream()
        continuation.finish()
    }

    private func shutdownSocketKeepingStream() {
        socket?.disconnect()
        socket?.removeAllHandlers()
        socket = nil
        manager = nil
        // Membership belongs to the connection we just dropped. Both
        // callers (login `connect`, logout `shutdown`) run with no
        // session view on screen, so there are no live holders to
        // orphan; keeping stale counts would instead make the next
        // `rejoinProjectRooms()` resubscribe rooms nobody is watching.
        projectRooms.removeAll()
    }

    // MARK: Rooms

    public func joinSession(_ id: String) { socket?.emit("subscribe:session", id) }
    public func leaveSession(_ id: String) { socket?.emit("unsubscribe:session", id) }

    /// Held project rooms → holder count. A REFCOUNT, not a flag: the
    /// inspector, the file-preview sheet, and anything else wanting
    /// `fs:changed` subscribe independently, and Socket.IO's `leave` is
    /// not refcounted — so without this the first holder to disappear
    /// unsubscribes the socket out from under the others.
    ///
    /// Doubles as the replay list for `rejoinProjectRooms()`.
    private(set) var projectRooms: [ProjectRoomKey: Int] = [:]

    public func joinProject(machineId: String, workingDir: String) {
        let key = ProjectRoomKey(machineId: machineId, workingDir: workingDir)
        let holders = (projectRooms[key] ?? 0) + 1
        projectRooms[key] = holders
        // Holders 2..N are already subscribed on this connection.
        if holders == 1 { emitJoinProject(key) }
    }

    public func leaveProject(machineId: String, workingDir: String) {
        let key = ProjectRoomKey(machineId: machineId, workingDir: workingDir)
        let holders = (projectRooms[key] ?? 0) - 1
        if holders > 0 {
            projectRooms[key] = holders
            return
        }
        projectRooms[key] = nil
        socket?.emit("unsubscribe:project", ["machineId": machineId, "workingDir": workingDir])
    }

    /// Re-subscribe every held project room after a reconnect.
    ///
    /// Socket.IO rooms are per-CONNECTION, so a drop silently discards
    /// all of them. Nothing replayed them before, so after any blip
    /// `fs:changed` / `git:changed` stopped arriving until the holding
    /// view happened to reappear — on mobile, where drops are routine,
    /// that meant the inspector's file tree and commit list quietly went
    /// dead while still looking live.
    ///
    /// Called from `AppModel`'s `.connected` handler, next to the
    /// session/terminal rejoin. Deliberately NOT called from the socket's
    /// own connect callback: those closures run outside this actor
    /// (SocketIO is a `@preconcurrency` import) and must not touch
    /// actor-isolated state — which is why every other handler in this
    /// file only ever touches `continuation`.
    public func rejoinProjectRooms() {
        for key in projectRooms.keys { emitJoinProject(key) }
    }

    private func emitJoinProject(_ key: ProjectRoomKey) {
        socket?.emit(
            "subscribe:project",
            ["machineId": key.machineId, "workingDir": key.workingDir]
        )
    }

    public func joinTerminal(_ id: String) { socket?.emit("subscribe:terminal", id) }
    public func leaveTerminal(_ id: String) { socket?.emit("unsubscribe:terminal", id) }

    // MARK: Terminal input (client → server; bytes ride base64 in JSON)

    public func sendTerminalInput(terminalId: String, base64Data: String) {
        socket?.emit("terminal:input", ["terminalId": terminalId, "data": base64Data])
    }

    public func sendTerminalResize(terminalId: String, cols: Int, rows: Int) {
        socket?.emit("terminal:resize", ["terminalId": terminalId, "cols": cols, "rows": rows])
    }

    public func sendTerminalClose(terminalId: String) {
        socket?.emit("terminal:close", terminalId)
    }

    // MARK: Handlers

    private func registerHandlers(on socket: SocketIOClient) {
        socket.on(clientEvent: .connect) { [weak self] _, _ in
            self?.continuation.yield(.connected)
        }
        socket.on(clientEvent: .disconnect) { [weak self] _, _ in
            self?.continuation.yield(.disconnected)
        }
        socket.on(clientEvent: .error) { [weak self] data, _ in
            self?.continuation.yield(.socketError(String(describing: data.first ?? "socket error")))
        }

        // Closure literals, not `ServerEvent.case` references: the
        // compiler doesn't infer @Sendable for unapplied enum-case
        // constructors, so the references warn under strict concurrency.
        on(socket, "chunk") { .chunk($0) }
        on(socket, "command:created") { .commandCreated($0) }
        on(socket, "command:updated") { .commandUpdated($0) }
        on(socket, "session:created") { .sessionCreated($0) }
        on(socket, "session:updated") { .sessionUpdated($0) }
        on(socket, "session:status") { .sessionStatus($0) }
        on(socket, "session:clone-failed") { .sessionCloneFailed($0) }
        on(socket, "machine:upsert") { .machineUpsert($0) }
        on(socket, "machine:status") { .machineStatus($0) }
        on(socket, "machine:removed") { .machineRemoved($0) }
        on(socket, "project:upsert") { .projectUpsert($0) }
        on(socket, "fs:changed") { .fsChanged($0) }
        on(socket, "git:changed") { .gitChanged($0) }
        on(socket, "background-task:updated") { .backgroundTaskUpdated($0) }
        on(socket, "background-task:removed") { .backgroundTaskRemoved($0) }
        on(socket, "terminal:created") { .terminalCreated($0) }
        on(socket, "terminal:updated") { .terminalUpdated($0) }
        on(socket, "terminal:output") { .terminalOutput($0) }
        on(socket, "terminal:closed") { .terminalClosed($0) }
    }

    /// Register a typed handler: decode the event's first argument into
    /// `T`, wrap with `make`, yield. Undecodable payloads are dropped —
    /// realtime events always have a REST fallback path.
    private func on<T: Decodable>(
        _ socket: SocketIOClient,
        _ event: String,
        _ make: @escaping @Sendable (T) -> ServerEvent
    ) {
        socket.on(event) { [weak self] data, _ in
            guard let value: T = Self.decodeFirst(data) else { return }
            self?.continuation.yield(make(value))
        }
    }

    /// Decode the first Socket.IO argument (a JSON object) by
    /// round-tripping JSONSerialization → JSONDecoder.
    private static func decodeFirst<T: Decodable>(_ data: [Any]) -> T? {
        guard let first = data.first,
              JSONSerialization.isValidJSONObject(first),
              let json = try? JSONSerialization.data(withJSONObject: first)
        else { return nil }
        return try? JSONDecoder().decode(T.self, from: json)
    }
}

File: apps/android/core/src/main/kotlin/app/argus/core/realtime/StreamClient.kt
package app.argus.core.realtime

import app.argus.core.ArgusJson
import app.argus.core.model.CommandDTO
import app.argus.core.model.MachineDTO
import app.argus.core.model.ProjectDTO
import app.argus.core.model.ResultChunk
import app.argus.core.model.SessionDTO
import app.argus.core.model.SessionStatusEvent
import app.argus.core.model.TerminalDTO
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow
import okhttp3.OkHttpClient
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI

/**
 * Realtime layer — the Kotlin counterpart of ArgusKit's `StreamClient`
 * and of `apps/web/src/lib/ws.ts`, against the event map in
 * `packages/shared-types/src/ws.ts`.
 *
 * Auth: the server (`stream.gateway.ts#handleConnection`) reads the JWT
 * from `handshake.auth.token`, sent here as the Socket.IO `auth` payload —
 * unlike a header this travels inside the Engine.IO CONNECT packet, so it
 * survives proxies.
 *
 * Delivery is best-effort: on [ServerEvent.Connected] after a drop, the
 * app must backfill missed chunks over REST (`getSessionChunks(afterSeq)`)
 * — there is no socket-side replay buffer — and replay its room
 * memberships with [rejoinProjectRooms] (rooms are per connection).
 *
 * Threading: Socket.IO invokes listeners on its own thread; they only
 * touch the unbounded [Channel], which is thread-safe. Everything else
 * (connect, rooms, emits) is expected from one thread — the app's main.
 * Create one per login; [connect] after login, [shutdown] on logout.
 */
class StreamClient(private val httpClient: OkHttpClient = OkHttpClient()) {
    private val channel = Channel<ServerEvent>(Channel.UNLIMITED)

    /** Consume once; delivers every event for the client's lifetime. */
    val events: Flow<ServerEvent> = channel.receiveAsFlow()

    private var socket: Socket? = null
    private val rooms = ProjectRoomRegistry()

    /** Held project rooms → holder count (see [ProjectRoomRegistry]). */
    val projectRooms: Map<ProjectRoomKey, Int>
        get() = rooms.snapshot()

    // MARK: Lifecycle

    fun connect(baseUrl: String, token: String) {
        shutdownSocketKeepingStream()
        // Reconnection mirrors ws.ts: always retry, capped backoff.
        val options = IO.Options.builder()
            .setTransports(arrayOf("websocket"))
            .setReconnection(true)
            .setReconnectionDelay(1_000)
            .setReconnectionDelayMax(10_000)
            .setAuth(mapOf("token" to token))
            .build()
        options.callFactory = httpClient
        options.webSocketFactory = httpClient
        // The URI path IS the namespace for socket.io-client-java.
        val socket = IO.socket(URI.create(baseUrl.trimEnd('/') + NAMESPACE), options)
        this.socket = socket
        registerHandlers(socket)
        socket.connect()
    }

    fun shutdown() {
        shutdownSocketKeepingStream()
        channel.close()
    }

    private fun shutdownSocketKeepingStream() {
        socket?.let {
            it.off()
            it.disconnect()
        }
        socket = null
        // Membership belongs to the connection we just dropped. Both
        // callers (login `connect`, logout `shutdown`) run with no session
        // view on screen, so there are no live holders to orphan; keeping
        // stale counts would instead make the next rejoin resubscribe
        // rooms nobody is watching.
        rooms.clear()
    }

    // MARK: Rooms

    fun joinSession(id: String) {
        socket?.emit("subscribe:session", id)
    }

    fun leaveSession(id: String) {
        socket?.emit("unsubscribe:session", id)
    }

    fun joinProject(machineId: String, workingDir: String) {
        val key = ProjectRoomKey(machineId, workingDir)
        // Holders 2..N are already subscribed on this connection.
        if (rooms.join(key)) emitJoinProject(key)
    }

    fun leaveProject(machineId: String, workingDir: String) {
        val key = ProjectRoomKey(machineId, workingDir)
        if (rooms.leave(key)) {
            socket?.emit("unsubscribe:project", projectPayload(key))
        }
    }

    /**
     * Re-subscribe every held project room after a reconnect. Socket.IO
     * rooms are per-CONNECTION, so a drop silently discards all of them;
     * without a replay, `fs:changed` / `git:changed` stop arriving until
     * the holding view happens to reappear. Call from the app's
     * [ServerEvent.Connected] handler, next to the session/terminal
     * rejoin. Does not mutate the refcounts — the holders haven't
     * changed, only the connection has.
     */
    fun rejoinProjectRooms() {
        for (key in rooms.keys()) emitJoinProject(key)
    }

    private fun emitJoinProject(key: ProjectRoomKey) {
        socket?.emit("subscribe:project", projectPayload(key))
    }

    private fun projectPayload(key: ProjectRoomKey): JSONObject =
        JSONObject().put("machineId", key.machineId).put("workingDir", key.workingDir)

    fun joinTerminal(id: String) {
        socket?.emit("subscribe:terminal", id)
    }

    fun leaveTerminal(id: String) {
        socket?.emit("unsubscribe:terminal", id)
    }

    // MARK: Terminal input (client → server; bytes ride base64 in JSON)

    fun sendTerminalInput(terminalId: String, base64Data: String) {
        socket?.emit("terminal:input", JSONObject().put("terminalId", terminalId).put("data", base64Data))
    }

    fun sendTerminalResize(terminalId: String, cols: Int, rows: Int) {
        socket?.emit(
            "terminal:resize",
            JSONObject().put("terminalId", terminalId).put("cols", cols).put("rows", rows),
        )
    }

    fun sendTerminalClose(terminalId: String) {
        socket?.emit("terminal:close", terminalId)
    }

    // MARK: Handlers

    private fun registerHandlers(socket: Socket) {
        socket.on(Socket.EVENT_CONNECT, Emitter.Listener { channel.trySend(ServerEvent.Connected) })
        socket.on(Socket.EVENT_DISCONNECT, Emitter.Listener { channel.trySend(ServerEvent.Disconnected) })
        socket.on(
            Socket.EVENT_CONNECT_ERROR,
            Emitter.Listener { args ->
                channel.trySend(ServerEvent.SocketError(args.firstOrNull()?.toString() ?: "socket error"))
            },
        )

        on<ResultChunk>(socket, "chunk") { ServerEvent.Chunk(it) }
        on<CommandDTO>(socket, "command:created") { ServerEvent.CommandCreated(it) }
        on<CommandDTO>(socket, "command:updated") { ServerEvent.CommandUpdated(it) }
        on<SessionDTO>(socket, "session:created") { ServerEvent.SessionCreated(it) }
        on<SessionDTO>(socket, "session:updated") { ServerEvent.SessionUpdated(it) }
        on<SessionStatusEvent>(socket, "session:status") { ServerEvent.SessionStatusChanged(it) }
        on<SessionCloneFailedPayload>(socket, "session:clone-failed") { ServerEvent.SessionCloneFailed(it) }
        on<MachineDTO>(socket, "machine:upsert") { ServerEvent.MachineUpsert(it) }
        on<IdStatusPayload>(socket, "machine:status") { ServerEvent.MachineStatusChanged(it) }
        on<IdPayload>(socket, "machine:removed") { ServerEvent.MachineRemoved(it) }
        on<ProjectDTO>(socket, "project:upsert") { ServerEvent.ProjectUpsert(it) }
        on<FSChangedPayload>(socket, "fs:changed") { ServerEvent.FsChanged(it) }
        on<GitChangedPayload>(socket, "git:changed") { ServerEvent.GitChanged(it) }
        on<TerminalDTO>(socket, "terminal:created") { ServerEvent.TerminalCreated(it) }
        on<TerminalDTO>(socket, "terminal:updated") { ServerEvent.TerminalUpdated(it) }
        on<TerminalOutputPayload>(socket, "terminal:output") { ServerEvent.TerminalOutput(it) }
        on<TerminalClosedPayload>(socket, "terminal:closed") { ServerEvent.TerminalClosed(it) }
    }

    /**
     * Register a typed handler: decode the event's first argument into
     * [T], wrap with [make], send. Undecodable payloads are dropped —
     * realtime events always have a REST fallback path.
     */
    private inline fun <reified T> on(socket: Socket, event: String, crossinline make: (T) -> ServerEvent) {
        socket.on(
            event,
            Emitter.Listener { args ->
                val first = args.firstOrNull() ?: return@Listener
                val value = decodeArg<T>(first) ?: return@Listener
                channel.trySend(make(value))
            },
        )
    }

    /**
     * socket.io-client-java hands objects over as org.json values; round-
     * trip through their JSON text into the tolerant ArgusJson decoder.
     */
    private inline fun <reified T> decodeArg(arg: Any): T? {
        val text = when (arg) {
            is JSONObject -> arg.toString()
            is JSONArray -> arg.toString()
            is String -> arg
            else -> return null
        }
        return runCatching { ArgusJson.decodeFromString<T>(text) }.getOrNull()
    }

    companion object {
        const val NAMESPACE = "/stream"
    }
}
